sql/src/main/java/io/crate/jobs/PageDownstreamContext.java

```diff
@@ -143,24 +122,28 @@ public void setBucket(int bucketIdx, Bucket rows, boolean isLast, PageResultList
                 exhausted.set(bucketIdx);
             }
             if (bucketsByIdx.size() == numBuckets) {
-                triggerConsumption();
+                mergeAndTriggerConsumer();
             }
         }
     }
 
-    private void triggerConsumption() {
-        mergeBuckets();
-        if (allUpstreamsExhausted()) {
-            pagingIterator.finish();
-        }
+    private void triggerConsumer(@Nullable Throwable t) {
         if (receivingFirstPage) {
-            consumer.accept(batchIterator, lastThrowable);
             receivingFirstPage = false;
+            consumer.accept(batchPagingIterator, t);
```

Comment:
as discussed, this should not be required, is there a reason you found for doing this?